### ISSUE TITLE
[Performance][DCU] FlagGems pointwise_dynamic adds high steady-state latency to aten::add.Tensor

### ISSUE THREAD

**Issue by Hsmall-lemon** [flagos-ai/FlagGems#6570]
## Issue Type

- [x] Performance Regression (Torch-FL / FlagGems operator integration)

## Summary

On Hygon DCU, Torch-FL routes `aten::add.Tensor` to `flagos_python` when mixed FlagGems is enabled. For an ESM-2 residual-add layout, the output is numerically correct, but the native FlagGems route takes **239.252 us** per synchronized call while the same operation through DAS boxing takes **25.335 us**: **9.443x** slower on the same physical device.

The supplied standalone reproducer then bypasses the Torch-FL ATen-to-Python bridge and calls `flag_gems.ops.add.add` / its `add_func` directly. The latency remains **211.666 us** and **207.011 us**, respectively. `add_func` is a `PointwiseDynamicFunction`, so this localizes the dominant fixed cost to FlagGems' generic `pointwise_dynamic` path, rather than ESM-2 model code or DAS boxing.

**Ownership:** Torch-FL selects the configured mixed route correctly (`add.Tensor = flagos_python`). The performance regression is in the generic FlagGems pointwise dispatch path used after that selection: repeated argument preparation, output/`StridedBuffer` preparation, generated-wrapper execution, and associated Python/Triton launch work. This is a performance issue, not a correctness failure.

## Environment (for performance reports)

- Hardware/platform: Hygon DCU / BW1000, physical device 0.
- Test node: `a14r1n04`.
- Hardware reproduction date: 2026-09-21.
- Runtime environment: `source /work1/share/scnethpc2601/flagos/activate.sh`.
- Python: `3.10.20`.
- PyTorch: `2.10.0+cpu` with HIP `6.3.26113`.
- Torch-FL: shared DCU installation; mixed config loaded from `backends_dcu_flaggems.conf`.
- FlagGems: shared FlagGems source used by that environment.
- Device visibility: `HIP_VISIBLE_DEVICES=0`, `CUDA_VISIBLE_DEVICES=0`.
- Input: FP32 ESM-style residual add, logical shape `(101, 3, 320)`. The left input is a transpose view with stride `(960, 320, 1)`.

The two routes were run in separate fresh Python processes on the same node and visible physical device, with identical data generation seed, dtype, shape, warmup count, repeat count, and synchronization policy.

## Reproduction

Save the following self-contained program as `repro_pointwise_dynamic_add.py`.

```python
import argparse
import os
import statistics
import time

parser = argparse.ArgumentParser()
parser.add_argument("--route", choices=("boxing", "flaggems"), required=True)
parser.add_argument("--warmup", type=int, default=30)
parser.add_argument("--repeat", type=int, default=300)
args = parser.parse_args()

if args.warmup < 1 or args.repeat < 1:
    raise ValueError("--warmup and --repeat must both be positive")

# Route selection must happen before import torch_fl. Torch-FL caches routing.
if args.route == "boxing":
    os.environ.pop("FLAGOS_USE_FLAGGEMS", None)
else:
    os.environ["FLAGOS_USE_FLAGGEMS"] = "1"
os.environ.pop("FLAGOS_USE_FLAGGEMS_CPP", None)
os.environ.pop("FLAGOS_BACKEND_CONFIG", None)

import torch
import torch_fl


def synchronize():
    torch.flagos.synchronize()


def synchronized_samples(fn, warmup, repeat):
    for _ in range(warmup):
        fn()
    synchronize()

    values_us = []
    output = None
    for _ in range(repeat):
        synchronize()
        started_ns = time.perf_counter_ns()
        output = fn()
        synchronize()
        values_us.append((time.perf_counter_ns() - started_ns) / 1_000.0)
    return statistics.median(values_us), output


def host_samples(fn, warmup, repeat):
    for _ in range(warmup):
        fn()
    values_us = []
    output = None
    for _ in range(repeat):
        started_ns = time.perf_counter_ns()
        output = fn()
        values_us.append((time.perf_counter_ns() - started_ns) / 1_000.0)
    return statistics.median(values_us), output


torch.manual_seed(20260921)
left_cpu = torch.randn(3, 101, 320).transpose(0, 1)
right_cpu = torch.randn(101, 3, 320)
expected = left_cpu + right_cpu
left = left_cpu.to("flagos:0")
right = right_cpu.to("flagos:0")

aten_us, result = synchronized_samples(
    lambda: torch.ops.aten.add.Tensor(left, right, alpha=1),
    args.warmup,
    args.repeat,
)
torch.testing.assert_close(result.cpu(), expected, rtol=1e-4, atol=1e-5)
assert result.stride() == left.stride()

print("route:", args.route)
print("backend_config:", os.environ.get("FLAGOS_BACKEND_CONFIG", "<unset>"))
print("correctness: PASS (CPU allclose; output stride preserved)")
print(f"aten_add_host_sync_median_us: {aten_us:.3f}")

if args.route == "boxing":
    raise SystemExit(0)

from flag_gems.ops.add import add as generic_add
from flag_gems.ops.add import add_func

# generic_add removes only the Torch-FL ATen-to-Python bridge. It retains
# FlagGems add semantics and pointwise_dynamic.
generic_us, generic_result = synchronized_samples(
    lambda: generic_add(left, right, alpha=1), args.warmup, args.repeat
)
torch.testing.assert_close(generic_result.cpu(), expected, rtol=1e-4, atol=1e-5)
assert generic_result.stride() == left.stride()

# add_func is the PointwiseDynamicFunction itself:
# prepare_args -> instantiate(cached after warmup) -> generated wrapper.
pointwise_us, pointwise_result = synchronized_samples(
    lambda: add_func(left, right, 1), args.warmup, args.repeat
)
torch.testing.assert_close(pointwise_result.cpu(), expected, rtol=1e-4, atol=1e-5)
assert pointwise_result.stride() == left.stride()

prepare_us, prepared = synchronized_samples(
    lambda: add_func.prepare_args(left, right, 1), args.warmup, args.repeat
)
ndim, prepared_args, prepared_kwargs = prepared
instantiate_us, overload = host_samples(
    lambda: add_func.instantiate(ndim), args.warmup, args.repeat
)

# These already-prepared arguments contain a reusable output buffer. This call
# deliberately excludes prepare_args and enters only the generated wrapper.
wrapper_us, _ = synchronized_samples(
    lambda: overload(*prepared_args, **prepared_kwargs), args.warmup, args.repeat
)
wrapper_result = prepared_kwargs["out0"].unwrap()
torch.testing.assert_close(wrapper_result.cpu(), expected, rtol=1e-4, atol=1e-5)
assert wrapper_result.stride() == left.stride()

print("generic_add_module:", generic_add.__module__)
print("pointwise_function_type:", type(add_func).__name__)
print(f"generic_add_host_sync_median_us: {generic_us:.3f}")
print(f"pointwise_dynamic_full_host_sync_median_us: {pointwise_us:.3f}")
print(f"prepare_args_host_sync_median_us: {prepare_us:.3f}")
print(f"instantiate_cached_host_median_us: {instantiate_us:.3f}")
print(f"generated_wrapper_host_sync_median_us: {wrapper_us:.3f}")
```

Run the two routes in separate processes after activating the same DCU environment:

```bash
source /work1/share/scnethpc2601/flagos/activate.sh
export ACCELERATOR=dcu
export TRITON_BACKENDS_IN_TREE=1
export HIP_VISIBLE_DEVICES=0
export CUDA_VISIBLE_DEVICES=0

unset FLAGOS_USE_FLAGGEMS FLAGOS_USE_FLAGGEMS_CPP FLAGOS_BACKEND_CONFIG
python repro_pointwise_dynamic_add.py --route boxing --warmup 30 --repeat 300

unset FLAGOS_USE_FLAGGEMS_CPP FLAGOS_BACKEND_CONFIG
export FLAGOS_USE_FLAGGEMS=1
python repro_pointwise_dynamic_add.py --route flaggems --warmup 30 --repeat 300
```

The script has no model, checkpoint, or project-file dependency. It creates its own inputs and validates each measured FlagGems result against CPU before reporting a number.

## Measured Output

Both routes completed with a CPU `allclose` pass (`rtol=1e-4`, `atol=1e-5`) and preserved the non-contiguous output stride.

```text
# DAS boxing
backend_config: .../backends_cuda.conf
aten_add_host_sync_median_us: 25.335

# Mixed FlagGems
backend_config: .../backends_dcu_flaggems.conf
aten_add_host_sync_median_us: 239.252
generic_add_module: flag_gems.ops.add
pointwise_function_type: PointwiseDynamicFunction
generic_add_host_sync_median_us: 211.666
pointwise_dynamic_full_host_sync_median_us: 207.011
prepare_args_host_sync_median_us: 46.400
instantiate_cached_host_median_us: 0.930
generated_wrapper_host_sync_median_us: 115.196
```

### Step 1: Localize the dominant layer from outside to inside

Each row below independently invokes the same `(101, 3, 320)` add from a different entry point, then measures it with synchronization. This is not a single additive time tree. Instead, each inner entry point deliberately bypasses one outer layer to determine whether the latency remains:

```text
full torch.add
  -> FlagGems generic_add without the Torch-FL bridge
    -> add_func / PointwiseDynamicFunction directly
```

| Measurement entry point | What this measurement deliberately bypasses | Median latency | What the row proves |
| --- | --- | ---: | --- |
| DAS boxing `aten::add.Tensor` | Nothing; baseline | `25.335 us` | The same add completes at approximately `25 us` through the normal control path. |
| Mixed FlagGems `aten::add.Tensor` | Nothing; full user-visible FlagGems path | `239.252 us` | FlagGems is `213.917 us` slower, or `9.443x` boxing. The regression is real. |
| FlagGems `generic_add` | Torch-FL ATen-to-Python bridge; retains generic FlagGems add | `211.666 us` | Removing the bridge saves only `27.586 us`; `211.666 us` remains far above `25.335 us`, so the bridge is not the dominant problem. |
| `PointwiseDynamicFunction` full call | Small outer logic in `generic_add`; directly calls `add_func` | `207.011 us` | Nearly all generic-add latency remains: `207.011 / 211.666 = 97.8%`. The dominant cost is inside `pointwise_dynamic`. |

In other words, direct `PointwiseDynamicFunction` execution still takes `207.011 us`, only `4.655 us` less than `generic_add`, yet it is about `181.676 us` slower than full DAS boxing. The optimization target is therefore the generic `pointwise_dynamic` execution path, not model code and not merely the outer Torch-FL bridge.

### Step 2: Identify expensive repeated work inside `pointwise_dynamic`

After locating the issue in `pointwise_dynamic`, the following internal entry points were measured separately. They answer which generic work is expensive on each call; they are not an additive partition of `207.011 us`:

| Internal `pointwise_dynamic` measurement | Work performed | Directly measured median | Interpretation |
| --- | --- | ---: | --- |
| `prepare_args` | Validation, dtype promotion, broadcast shape/stride analysis, output allocation, and `StridedBuffer` wrapping | `46.400 us` | Preparing one generic pointwise add alone is `1.83x` the full boxing add (`25.335 us`). This fixed work needs caching or elimination for repeated signatures. |
| Cached `instantiate` | Lookup of the already-generated rank-specialized overload | `0.930 us` | Negligible. Steady-state latency is not repeated Python source generation or Triton compilation. |
| Generated wrapper with prepared arguments | Skips `prepare_args`; executes the generated Triton wrapper, generic launch preparation, and device submission | `115.196 us` | Even with metadata already prepared, this generic wrapper path is `4.55x` the full boxing call. The wrapper, argument-packaging, and launch path also need optimization. |

`46.400 us` and `115.196 us` are both direct diagnostic measurements, but they must not be added to form `161.596 us`, nor subtracted from `207.011 us` and called a kernel duration: the tests use different prepared-state, output-allocation, and synchronization conditions. What the data rigorously shows is that the complete steady-state `pointwise_dynamic` path costs `207.011 us`, its argument preparation and generated-wrapper paths are independently large and optimizable, and cached overload lookup is not the bottleneck. This reproduction intentionally makes no unverified device-kernel-time claim.

## Root Cause Analysis / Bottom-layer Execution

The observed execution path is:

```text
torch.ops.aten.add.Tensor(left, right, alpha=1)
  -> Torch-FL config: add.Tensor = flagos_python
  -> Torch-FL generated Python callable bridge
  -> flag_gems.ops.add.add(left, right, alpha=1)
  -> add_func(left, right, alpha)
  -> PointwiseDynamicFunction.__call__
  -> prepare_args(...)
  -> instantiate(rank)                    # cached after warmup
  -> generated rank-3 Triton wrapper
  -> DCU kernel launch
```

The outside-to-inside experiment provides a clear localization: full FlagGems costs `239.252 us`; bypassing the Torch-FL bridge still costs `211.666 us`; and direct `PointwiseDynamicFunction` invocation still costs `207.011 us`. The dominant problem is therefore `pointwise_dynamic`, rather than model code or the outer Torch-FL bridge. `pointwise_dynamic.prepare_args` is not only a lookup. For a generic pointwise operator it validates tensor/scalar positions, applies dtype-promotion rules, identifies the broadcasted task shape, analyzes strides and layout, allocates the output when needed, and wraps inputs/outputs as `StridedBuffer` objects. The generated wrapper then does generic argument packing and launch preparation for the current rank/layout. Those semantics are needed for arbitrary `add` inputs, but repeatedly executing all of them imposes a high fixed cost on short ESM residual additions.

`instantiate` is already cached after warmup and takes only `0.930 us`. The optimization should therefore reduce repeated metadata preparation and shorten the generic generated-wrapper launch path, not focus on avoiding repeated compilation.

## Expected Behavior

FlagGems must retain complete PyTorch `add` semantics for arbitrary dtype, broadcast, rank, and layout cases. For steady-state common eager inputs, however, native FlagGems execution should not be nearly an order of magnitude slower than DAS boxing solely because generic per-call dispatch work is repeated.

The acceptance criterion is not a model-side workaround or an environment override. The native `flagos_python`/FlagGems path should keep numerical and stride correctness while substantially reducing the fixed pointwise dispatch latency.

## Proposed Solution / Upstream Regression Plan

1. Add a reusable fast dispatch layer inside FlagGems for repeated pointwise signatures keyed by operation, dtype, rank, shape, stride, broadcast pattern, scalar attributes, and output layout. It must preserve the generic path for signatures that are not safely covered.
2. Avoid reconstructing equivalent metadata objects on every steady-state call when the input signature is unchanged. In particular, assess caching or a lower-level representation for broadcasted strides, task shape, output-layout decisions, and the generated-wrapper launch descriptor.
3. Reduce Python-object and Python/Triton wrapper work on common eager binary pointwise calls. A native C++/HIP launch path is acceptable only if it remains part of FlagGems and preserves the exact ATen semantics; it must not silently route the operator to DAS boxing.
4. Keep `instantiate` caching and add a regression that distinguishes first-use compilation from steady-state latency. The measured problem remains after the overload is cached.
5. Add the self-contained `(101, 3, 320)` transpose-view regression above, plus contiguous and broadcast variants. Compare native FlagGems with DAS boxing on the same device, validate CPU numerical output and output stride, and report the ratio without making a hardware-independent fixed threshold mandatory.

## Verification Plan

- [ ] Run the exact reproducer on the native mixed FlagGems route without per-operator CUDA overrides.
- [ ] Verify output against CPU (`rtol=1e-4`, `atol=1e-5`) and preserve the left operand's stride.
- [ ] Verify the generated overload is warm/cached before steady-state measurement.
- [ ] Compare boxing and FlagGems in separate fresh processes on the same physical DCU.
- [ ] Test the ESM residual layout, contiguous same-shape add, and valid broadcast inputs.
- [ ] Confirm the native FlagGems route materially reduces the fixed latency without changing model code.

## Checklist

- [x] Current DCU hardware reproduction recorded.
- [x] Self-contained runnable program included in this report.
- [x] DAS boxing control and native mixed FlagGems route compared on one device.
- [x] Numerical value and output-stride checks included.
- [x] `pointwise_dynamic` stages measured after warmup.
- [x] Cached `instantiate` distinguished from first-use compilation.
- [x] No private output path is required to understand or reproduce the issue.
